### PR TITLE
Add inkstitch svg version tag

### PR DESCRIPTION
**Pull request by kaalleen**
This adds a inkstitch version tag to the svg file to make legacy updates a bit easier.

I'm not very happy about when this is called. Any better idea?
(In future we may need to call it on metadata or def tags or whatever possible in the svg)

### CODE REVIEW

File: lib/extensions/base.py [hunk 1]
```diff
@@ -223,3 +227,45 @@ def get_base_file_name(self):
...
+            # let's udowngrade the file version number
+            self._update_inkstitch_svg_version()
+        else:
+            # this document is either a new document or it is outdated
+            # if we cannot find any inkstitch attribute in the document, we assume that this is a new document which doesn't need to be updated
+            search_string = "//*[namespace-uri()='http://inkstitch.org/namespace' or " \
+                            "@*[namespace-uri()='http://inkstitch.org/namespace'] or " \
+                            "@*[starts-with(name(), 'embroider_')]]"
+            inkstitch_element = svg.findone(search_string)
+            if inkstitch_element is None:
```
Comment (lexelby) [inline lib/extensions/base.py:258]: I think we want to update if there _are_ existing Ink/Stitch attributes in the doc, right? So this should be `inkstitch_element is not None`?

**Review by lexelby — COMMENTED** [review]
This is awesome, thanks for working on this!  I'm glad we're getting all of the funky legacy logic into one file, it's nice to remove the clutter.  I would be a fan of moving all of this, even `_update_inkstitch_document()`, into update.py too.

I think you're right about where that's called.  I'm gonna suggest something that I'm not quite sure about myself, so tell me if you think it's crazy.  I am thinking of overriding the `load()` method in InkstitchExtension.  We could do something like this:

```python
def load(self, *args, **kwargs):
    document = super().load(*args, **kwargs)
    update_inkstitch_document(document)
    return document
```

`load()` is normally called by `InkscapeExtension.load_raw()` which does `self.load(...)`.  If we override `load()` in our own class, then `load_raw()` will call _ours_, and then we call the parent class one to do the actual SVG loading.  Then we make our legacy changes and return it, and everything continues like normal but with an updated SVG document.

File: lib/extensions/base.py [hunk 2]
```diff
@@ -223,3 +227,45 @@ def get_base_file_name(self):
...
+
+    def _update_inkstitch_document(self, nodes):
+        # get the inkstitch svg version from the document
+        svg = self.document.getroot()
+        search_string = "//*[local-name()='inkstitch_svg_version']//text()"
+        file_version = svg.findone(search_string)
+        if file_version is None:
+            file_version = 0
+        else:
+            file_version = int(file_version)
```
[the remark this answers is not in the document]
Comment (kaalleen) [reply lib/extensions/base.py:239]: Much better, yes

File: lib/extensions/base.py [hunk 3]
```diff
@@ -223,3 +227,45 @@ def get_base_file_name(self):
...
+            inkstitch_element = svg.findone(search_string)
+            if inkstitch_element is None:
+                self._update_inkstitch_svg_version()
+                return
+
+            # update elements
+            elements = nodes_to_elements(nodes)
+            for element in elements:
+                update_legacy_params(element, file_version, INKSTITCH_SVG_VERSION)
+            self._set_inkstitch_svg_version()
```
[the remark this answers is not in the document]
Comment (kaalleen) [reply lib/extensions/base.py:266]: Yes, right.

File: lib/extensions/base.py [hunk 4]
```diff
@@ -223,3 +227,45 @@ def get_base_file_name(self):
     def uniqueId(self, prefix, make_new_id=True):
         """Override inkex.Effect.uniqueId with a nicer naming scheme."""
         return generate_unique_id(self.document, prefix)
+
+    def _update_inkstitch_document(self, nodes):
+        # get the inkstitch svg version from the document
+        svg = self.document.getroot()
+        search_string = "//*[local-name()='inkstitch_svg_version']//text()"
```
[the remark this answers is not in the document]
Comment (kaalleen) [reply lib/extensions/base.py:234]: If we did it like that, we would already set inkstitch namespaced attributes and we cannot check for new documents later on.

File: lib/extensions/base.py (hunk 1 above)
Comment (kaalleen) [reply lib/extensions/base.py:258]: Yes, but we still need to set the inkstitch svg file version if we do not want to run the update check again.

**Comment by kaalleen** [comment]
Thank you for your review. Especially the code example for the load method was very helpful.
I actually did think about this method but didn't figure out how to make it work in there.

**Pull request merged**